Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalWaferP.cc

```diff
@@ -0,0 +1,193 @@
+/*
+ * DDHGCalWaferP.cc
+ *
+ *  Created on: 09-Jan-2021
+ */
+
+#include "DD4hep/DetFactoryHelper.h"
+#include "DetectorDescription/DDCMS/interface/DDPlugins.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "Geometry/HGCalCommonData/interface/HGCalTypes.h"
+#include "Geometry/HGCalCommonData/interface/HGCalWaferMask.h"
+
+#include <string>
+#include <vector>
+#include <sstream>
+
+//#define EDM_ML_DEBUG
+
+static long algorithm(dd4hep::Detector& /* description */, cms::DDParsingContext& ctxt, xml_h e) {
+#ifdef EDM_ML_DEBUG
+  static constexpr double f2mm = (1.0 / dd4hep::mm);
+#endif
+  cms::DDNamespace ns(ctxt, e, true);
+  cms::DDAlgoArguments args(ctxt, e);
+  std::string parentName = args.parentName();
+  const auto& material = args.value<std::string>("ModuleMaterial");
+  const auto& thick = args.value<double>("ModuleThickness");
+  const auto& waferSize = args.value<double>("WaferSize");
+  const auto& waferSepar = args.value<double>("SensorSeparation");
+#ifdef EDM_ML_DEBUG
+  edm::LogVerbatim("HGCalGeom") << "DDHGCalWaferP: Module " << parentName << " made of " << material << " T "
+                                << (f2mm * thick) << " Wafer 2r " << (f2mm * waferSize) << " Half Separation "
```

Comment:
@bsunanda Can you explain why this code is converting cm to mm only in the debug messages, but the code itself is still doing the calculations in cm? However, the DDD version of this file is doing the same calculations in mm. What am I missing?

Reply:
I compare the debug printout from DDD and DD4Hep. In DDD these are in mm - so I print out in mm for dd4hep. I do not touch the interface to dd4hep for any of the units.

Reply:
OK. These algorithms are completely internal to DD4hep. Currently, the calculations are done in cm. When we switch to the new DD4hep, they will be done internally in mm. Since the algorithm code does not contain units assumptions (except in the debug output, which are handled by the conversion factor), it will work the same in either case.